src/service/resources-impl.js

```diff
@@ -468,80 +483,105 @@ export class ResourcesImpl {
       this.ampdoc.getVisibilityState() != VisibilityState.PRERENDER ||
       resource.prerenderAllowed();
 
-    if (buildingEnabled && shouldBuildResource) {
+    const isCloseEnoughToViewport = this.buildWhenCloseToViewport_
+      ? ignoreQuota ||
+        resource.isBuildRenderBlocking() ||
+        resource.renderOutsideViewport() ||
+        (this.isIdle_() && resource.idleRenderOutsideViewport())
+      : true;
+
+    if (
+      buildingEnabled &&
+      !isBuiltOrBuilding &&
+      shouldBuildResource &&
+      isCloseEnoughToViewport
+    ) {
       if (this.documentReady_) {
         // Build resource immediately, the document has already been parsed.
         this.buildResourceUnsafe_(resource, ignoreQuota);
-      } else if (!resource.isBuilt() && !resource.isBuilding()) {
-        if (!checkForDupes || !this.pendingBuildResources_.includes(resource)) {
-          // Otherwise add to pending resources and try to build any ready ones.
-          this.pendingBuildResources_.push(resource);
-          this.buildReadyResources_();
-        }
+      } else if (!this.pendingBuildResources_.includes(resource)) {
+        // Otherwise add to pending resources and try to build any ready ones.
+        this.pendingBuildResources_.push(resource);
+        this.buildReadyResources_(ignoreQuota);
       }
     }
   }
 
   /**
    * Builds resources that are ready to be built.
+   * @param {boolean} ignoreQuota
    * @private
    */
-  buildReadyResources_() {
+  buildReadyResources_(ignoreQuota) {
     // Avoid cases where elements add more elements inside of them
     // and cause an infinite loop of building - see #3354 for details.
     if (this.isCurrentlyBuildingPendingResources_) {
       return;
     }
     try {
       this.isCurrentlyBuildingPendingResources_ = true;
-      this.buildReadyResourcesUnsafe_();
+      this.buildReadyResourcesUnsafe_(ignoreQuota);
     } finally {
       this.isCurrentlyBuildingPendingResources_ = false;
     }
   }
 
   /**
+   * @param {boolean} ignoreQuota
    * @private
    */
-  buildReadyResourcesUnsafe_() {
+  buildReadyResourcesUnsafe_(ignoreQuota) {
     // This will loop over all current pending resources and those that
     // get added by other resources build-cycle, this will make sure all
     // elements get a chance to be built.
-    for (let i = 0; i < this.pendingBuildResources_.length; i++) {
-      const resource = this.pendingBuildResources_[i];
+    const pending = this.pendingBuildResources_;
+    const last = pending.length - 1;
+    let insertion = this.pendingBuildIndex_;
+    for (let i = insertion; i < pending.length; i++) {
+      const resource = pending[i];
+      if (resource === null) {
+        continue;
+      }
```

Comment:
Avoiding the `splice`? Don't think this is worth the readability cost. How many times is this function called?